### PR TITLE
feat: allow catching all exceptions

### PR DESCRIPTION
**Pull request by KevinEady**
Add new compile-time define (`NODE_ADDON_API_CPP_EXCEPTIONS_ALL`)  to enable catching all exceptions. This must be used _in conjunction with_ `NODE_ADDON_API_CPP_EXCEPTIONS`.

Rename the `NAPI_` prefixed exception preprocessor definitions (`NAPI_CPP_EXCEPTIONS` and `NAPI_DISABLE_CPP_EXCEPTIONS`) to be prefixed with our standard `NODE_ADDON_API_` prefix in a backwards-compatible way.

Fixes: #1555

### CODE REVIEW

**Comment by KevinEady** [comment]
I'd like to discuss the naming of the new define `NAPI_CPP_EXCEPTIONS_ALL` ... to go with convention it should be `NODE_ADDON_API_CPP_EXCEPTIONS_ALL`.

We can use `NODE_ADDON_API_CPP_EXCEPTIONS_ALL` and introduce `NODE_ADDON_API_CPP_EXCEPTIONS` as a (backwards-compatible) replacement for `NAPI_CPP_EXCEPTIONS_ALL` ?

**Comment by KevinEady** [comment]
We discuss in the 18 Oct Node-API meeting:
- Use the `NODE_ADDON_API_` prefix on the defines
- Should the "catch all" be the new default?
  - What about some best practices doc? We also mentioned this with the requiring basic finalizers option

File: napi-inl.h [hunk 1]
```diff
@@ -78,43 +78,61 @@ inline napi_status AttachData(napi_env env,
 // For use in JS to C++ callback wrappers to catch any Napi::Error exceptions
 // and rethrow them as JavaScript exceptions before returning from the callback.
 template <typename Callable>
-inline napi_value WrapCallback(Callable callback) {
-#ifdef NAPI_CPP_EXCEPTIONS
+inline napi_value WrapCallback(napi_env env, Callable callback) {
```
Comment (NickNaso) [inline napi-inl.h:81]: Why are adding the `env` parameter?

**Review by NickNaso — COMMENTED** [review]
Is this branch in sync with the `main`?

**Comment by KevinEady** [comment]
The CIs are failing either because (1) unused parameters, or (2) experimental mode basic_env type mismatches.

Regarding 2: Catching all exceptions will construct a new`Napi::Error` to wrap the native exception. This cannot be done on a `basic_env`:

https://github.com/nodejs/node-addon-api/blob/98aae3343c3af36b4befd6b67c4cb19ba49b8d20/napi-inl.h#L209-L227

This `Wrapper` is used when eg. constructing a new `Napi::External` for the finalizer. We can't pass the basic env to `WrapVoidCallback` since it needs a regular env to wrap the error.

This means that a combination of `NAPI_HAS_POST_FINALIZER` and `NODE_ADDON_API_CPP_EXCEPTIONS_ALL` cannot happen, since `Napi::Error`-wrapping cannot occur. 

Should we make it a compile-time `#error` if these two options are used in conjunction?

File: napi-inl.h [hunk 2]
```diff
@@ -112,10 +126,40 @@ inline void WrapVoidCallback(Callable callback) {
...
+  catch (const std::exception& e) {
+    Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
+  } catch (...) {
+    Napi::Error::New(env, "A native exception was thrown")
+        .ThrowAsJavaScriptException();
+  }
+#endif  // NODE_ADDON_API_CPP_EXCEPTIONS_ALL
+#else
+  // When C++ exceptions are disabled, errors are immediately thrown as JS
+  // exceptions, so there is no need to catch and rethrow them here.
```
Comment (legendecas) [inline napi-inl.h:154]: I think the term "thrown" should be clarified here. Technically, the JS errors are not "thrown" in C++ natively.

```suggestion
  // When C++ exceptions are disabled, there is no need to catch and rethrow C++ exceptions. JS errors should be thrown with `Error::ThrowAsJavaScriptException`.
```

**Review by legendecas — COMMENTED** [review]
Looks good to me! A minor suggestion...

**Review by legendecas — APPROVED** [review]

**Review by mhdawson — APPROVED** [review]
LGTM

**Pull request merged**